Update error messages to indicate the correct recipe-script to use when running standardized script on wrong backend

**Description of Problem**

See [my comment in #6](https://github.com/JordanMartinez/purescript-cookbook/pull/6#discussion_r438515582) and [my other comment]() for more context. We're currently using a standardized npm script to run a specific recipe using the npm config value. For example
```
npm run node --cookbook:recipe=MyRecipeName
npm run node --cookbook:recipe=BrowserExample
npm run browser --cookbook:recipe=HalogenExample
```
The problem arises when we use this standardized script to run a browser-only recipe. For example
```
npm run node --cookbook:recipe=HalogenApp
```
The error message should say "This recipe doesn't work on the Node.js backend. It only works on the browser backend. Please run `<correct command>` instead." (or vice versa for Node-only recipes)

Unfortunately, we currently get this:
```
$ npm run run --cookbook:recipe=RoutingHashHalogen

> cookbook@ run /home/jordan/Programming/Git Projects/purescript-cookbook
> spago run --main $npm_package_config_recipe.Main

[info] Installation complete.

[info] Build succeeded.
/home/jordan/Programming/Git Projects/purescript-cookbook/output/Effect.Aff/foreign.js:532
                throw util.fromLeft(step);
                ^

ReferenceError: window is not defined
    at exports.window (/home/jordan/Programming/Git Projects/purescript-cookbook/output/Web.HTML/foreign.js:5:3)
    at /home/jordan/Programming/Git Projects/purescript-cookbook/output/Effect/foreign.js:12:16
    at /home/jordan/Programming/Git Projects/purescript-cookbook/output/Effect/foreign.js:12:16
    at __do (/home/jordan/Programming/Git Projects/purescript-cookbook/output/Halogen.Aff.Util/index.js:36:183)
    at runAsync (/home/jordan/Programming/Git Projects/purescript-cookbook/output/Effect.Aff/foreign.js:98:20)
    at run (/home/jordan/Programming/Git Projects/purescript-cookbook/output/Effect.Aff/foreign.js:333:22)
    at /home/jordan/Programming/Git Projects/purescript-cookbook/output/Effect.Aff/foreign.js:639:15
    at drain (/home/jordan/Programming/Git Projects/purescript-cookbook/output/Effect.Aff/foreign.js:120:9)
    at Object.enqueue (/home/jordan/Programming/Git Projects/purescript-cookbook/output/Effect.Aff/foreign.js:141:11)
    at Object.run (/home/jordan/Programming/Git Projects/purescript-cookbook/output/Effect.Aff/foreign.js:638:23)
[error] Running failed; exit code: 1
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! cookbook@ run: `spago run --main $npm_package_config_recipe.Main`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the cookbook@ run script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.
npm WARN Local package.json exists, but node_modules missing, did you mean to install?

npm ERR! A complete log of this run can be found in:
npm ERR!     /home/jordan/.npm/_logs/2020-06-11T01_41_20_388Z-debug.log
```

--- a/.travis.yml
+++ b/.travis.yml
@@ -7,8 +7,9 @@ branches:
 install:
   - npm install
 script:
-  - diff README.md <(scripts/generateRecipeTable.sh)
   - make testAllCommands
+  - echo "\n\n\n"
+  - diff README.md <(scripts/generateRecipeTable.sh)
   - make testAllCI
 cache:
   directories:

--- a/.travis.yml
+++ b/.travis.yml
@@ -8,8 +8,8 @@ install:
   - npm install
 script:
   - diff README.md <(scripts/generateRecipeTable.sh)
-  - make testAllCI
   - make testAllCommands
+  - make testAllCI
 cache:
   directories:
    - output

--- a/.travis.yml
+++ b/.travis.yml
@@ -8,7 +8,6 @@ install:
   - npm install
 script:
   - make testAllCommands
-  - echo "\n\n\n"
   - diff README.md <(scripts/generateRecipeTable.sh)
   - make testAllCI
 cache:

--- a/.travis.yml
+++ b/.travis.yml
@@ -8,9 +8,8 @@ install:
   - npm install
 script:
   - diff README.md <(scripts/generateRecipeTable.sh)
-  - make buildAll
-  - make runAllNode
-  - make buildAllWeb
+  - make testAllCI
+  - make testAllCommands
 cache:
   directories:
    - output

--- a/makefile
+++ b/makefile
@@ -36,6 +36,14 @@ MAKEFLAGS += --no-builtin-rules
 
 # ===== Makefile - Repo Commands =====
 
+# Variables get that a list of all recipes that can be run on
+# Node.js or browser backends
+#
+#   wildcard = expand globs and add a space in-between each one
+#   patsubst = patsubst(textToFind,replaceItWithThisText,originalText)
+targetsNode := $(patsubst recipes/%/nodeSupported.md,%-node,$(wildcard recipes/*/nodeSupported.md))
+targetsWeb := $(patsubst recipes/%/web,%-web,$(wildcard recipes/*/web))
+
 # Note: usages of `.PHONY: <target>` mean the target name is not an actual file
 # .PHONY: targetName
 

--- a/makefile
+++ b/makefile
@@ -142,19 +142,22 @@ recipes/%/web:
 %-buildWeb: $(call recipeDir,%) $(call webDir,%) %-build
 > parcel build $(call webHtml,$*) --out-dir $(call webDistDir,$*) --no-minify --no-source-maps
 
+# Note: `%-buildProd` and its related commands were commented out
+# because the command doesn't currently work.
+
 # How to make prodDir
-recipes/%/prod:
-> mkdir -p $@
+# recipes/%/prod:
+# > mkdir -p $@
 
 # How to make prodHtml
-recipes/%/prod/index.html: $(call prodDir,%)
-> cp $(call webHtml,$*) $(call prodDir,$*)
+# recipes/%/prod/index.html: $(call prodDir,%)
+# > cp $(call webHtml,$*) $(call prodDir,$*)
 
 # Creates a minified production build.
 # For reference.
-%-buildProd: $(call recipeDir,%) $(call webDir,%) $(call prodHtml,%)
-> spago -x $(call recipeSpago,$*) bundle-app --main $(call main,$*) --to $(call prodJs,$*)
-> parcel build $(call prodHtml,$*) --out-dir $(call prodDistDir,$*)
+# %-buildProd: $(call recipeDir,%) $(call webDir,%) $(call prodHtml,%)
+# > spago -x $(call recipeSpago,$*) bundle-app --main $(call main,$*) --to $(call prodJs,$*)
+# > parcel build $(call prodHtml,$*) --out-dir $(call prodDistDir,$*)
 
 # ===== Makefile - CI Commands =====
 
@@ -192,4 +195,4 @@ testAllCommands:
 > $(MAKE) HelloWorld-build
 > $(MAKE) HelloWorld-buildWeb
 > $(MAKE) HelloWorld-testCI
-> $(MAKE) HelloWorld-buildProd
+# > $(MAKE) HelloWorld-buildProd

--- a/makefile
+++ b/makefile
@@ -9,6 +9,10 @@ ifeq ($(origin .RECIPEPREFIX), undefined)
 endif
 .RECIPEPREFIX = >
 
+# Use one shell to run all commands in a given target rather than using
+# the default setting of running each command in a separate shell
+.ONESHELL:
+
 # Make's default shell is `sh`. Setting this to `bash` enables
 # the shell flags below to work.
 SHELL := bash
@@ -70,34 +74,40 @@ info:
 
 # Tests if recipe actually exists.
 recipes/%:
-> @test -d $* || { echo "Recipe $* does not exist"; exit 1;}
+> @if [ ! -d $* ]
+> then
+>   echo
+>   echo Recipe $* does not exist
+>   echo
+>   exit 1
+> fi
 
-# Test if recipe is compatible with node
-recipes/%/nodeSupported.md:
-> @test -f $* || { echo "Recipe $* is not compatible with Node.js backend"; exit 1;}
-
-# Test if recipe is compatible with web
-recipes/%/web:
-> @test -f $* || { echo "Recipe $* is not compatible with web browser backend"; exit 1;}
-
-# Variables for all recipes' operations
-
+# Variables shared across most targets
 recipes := $(shell ls recipes)
 
-targetsNode := $(patsubst recipes/%/nodeSupported.md,%-node,$(wildcard recipes/*/nodeSupported.md))
-recipesWeb := $(patsubst recipes/%/web,%,$(wildcard recipes/*/web))
-
-targetsBuild := $(foreach r,$(recipes),$(r)-build)
-targetsWeb := $(foreach r,$(recipesWeb),$(r)-web)
-targetsBuildWeb := $(foreach r,$(recipesWeb),$(r)-buildWeb)
-targetsBuildProd := $(foreach r,$(recipesWeb),$(r)-buildProd)
-
-# Helper functions for generating paths
+# Functions shared across most targets that help generate paths
 main = $1.Main
 recipeDir = recipes/$1
 recipeSpago = $(call recipeDir,$1)/spago.dhall
 
+# Functions for Node.js-comptabile recipes that help generate paths
 nodeCompat = $(call recipeDir,$1)/nodeSupported.md
+
+# Tests whether recipe can be run on Node.js backend
+recipes/%/nodeSupported.md:
+> @if [ ! -f $(call nodeCompat,$*) ]
+> then
+>   echo
+>   echo Recipe $* is not compatible with Node.js backend
+>   echo
+>   exit 1
+> fi
+
+# Runs recipe as node.js console app
+%-node: $(call recipeDir,%) $(call nodeCompat,%)
+> spago -x $(call recipeSpago,$*) run --main $(call main,$*)
+
+# Functions for browser-comptabile recipes that help generate paths
 
 webDir = $(call recipeDir,$1)/web
 webHtml = $(call webDir,$1)/index.html
@@ -112,9 +122,15 @@ prodDistDir = $(call recipeDir,$1)/prod-dist
 %-build:
 > spago -x $(call recipeSpago,$*) build
 
-# Runs recipe as node.js console app
-%-node: $(call recipeDir,%) $(call nodeCompat,%)
-> spago -x $(call recipeSpago,$*) run --main $(call main,$*)
+# Tests whether recipe can be run on web browser backend
+recipes/%/web:
+> @if [ ! -d $(call webDir,$*) ]
+> then
+>   echo
+>   echo Recipe $* is not compatible with the web browser backend
+>   echo
+>   exit 1
+> fi
 
 # Launches recipe in web browser
 %-web: $(call recipeDir,%) $(call webDir,%) %-build
@@ -127,7 +143,7 @@ prodDistDir = $(call recipeDir,$1)/prod-dist
 > parcel build $(call webHtml,$*) --out-dir $(call webDistDir,$*) --no-minify --no-source-maps
 
 # How to make prodDir
-$(call prodDir,$(recipes)):
+recipes/%/prod:
 > mkdir -p $@
 
 # How to make prodHtml
@@ -136,22 +152,44 @@ recipes/%/prod/index.html: $(call prodDir,%)
 
 # Creates a minified production build.
 # For reference.
-%-buildProd: $(call recipeDir,%) $(call prodHtml,%)
+%-buildProd: $(call recipeDir,%) $(call webDir,%) $(call prodHtml,%)
 > spago -x $(call recipeSpago,$*) bundle-app --main $(call main,$*) --to $(call prodJs,$*)
 > parcel build $(call prodHtml,$*) --out-dir $(call prodDistDir,$*)
 
 # ===== Makefile - CI Commands =====
 
-.PHONY: buildAll runAllNode buildAllWeb buildAllProd
+# Runs a single recipe's CI.
+%-testCI: $(call recipeDir,%)
+> @if [ -f $(call nodeCompat,$*) ]
+> then
+>   echo Testing $* on the Node.js backend
+>   spago -x $(call recipeSpago,$*) run --main $(call main,$*)
+>   echo
+>   echo == $* - Succeeded on Node.js
+>   echo
+> fi
+> @if [ -d $(call webDir,$*) ]
+> then
+>   echo Attempting to build $* for the browser backend
+>   export NODE_ENV=development
+>   parcel build $(call webHtml,$*) --out-dir $(call webDistDir,$*) --no-minify --no-source-maps
+>   echo
+>   echo == $* - Built for browser
+>   echo
+> fi
 
-# All purs builds - for CI
-buildAll: $(targetsBuild)
+targetsAllCI := $(foreach r,$(recipes),$(r)-testCI)
 
-# All node executions - for CI
-runAllNode: $(targetsNode)
+# Run all recipes CI
+.PHONY: testAllCI
+testAllCI: $(targetsAllCI)
 
-# All web builds - for CI
-buildAllWeb: $(targetsBuildWeb)
-
-# All prod builds - for CI
-buildAllProd: $(targetsBuildProd)
+# Verifies that running all the above commands for a single
+# recipe actually works.
+testAllCommands:
+> $(MAKE)
+> $(MAKE) HelloWorld-node
+> $(MAKE) HelloWorld-build
+> $(MAKE) HelloWorld-buildWeb
+> $(MAKE) HelloWorld-testCI
+> $(MAKE) HelloWorld-buildProd

